environment/config.go

```diff
@@ -0,0 +1,231 @@
+)
+
+// Configuration is a configuration that can be used to create namespaces, a topology, and kv store
+type Configuration struct {
+	// Service is used when a topology initializer is not supplied.
+	Service *etcdclient.Configuration `yaml:"service"`
+
+	// StaticConfiguration is used for running M3DB with a static config
+	Static *StaticConfiguration `yaml:"static"`
+}
+
+// StaticConfiguration is used for running M3DB with a static config
+type StaticConfiguration struct {
+	Namespaces     []StaticNamespaceConfiguration `yaml:"namespaces"`
+	TopologyConfig *topology.StaticConfiguration  `yaml:"topology"`
+}
+
+// StaticNamespaceConfiguration sets the static namespace
+type StaticNamespaceConfiguration struct {
+	Name      string                    `yaml:"name"`
+	Options   *StaticNamespaceOptions   `yaml:"options"`
+	Retention *StaticNamespaceRetention `yaml:"retention"`
+}
+
+// StaticNamespaceOptions sets namespace options- if nil, default is used
+type StaticNamespaceOptions struct {
+	NeedsBootstrap      bool `yaml:"needsBootstrap"`
+	NeedsFlush          bool `yaml:"needsFlush"`
+	WritesToCommitLog   bool `yaml:"writesToCommitLog"`
+	NeedsFilesetCleanup bool `yaml:"needsFilesetCleanup"`
+	NeedsRepair         bool `yaml:"needsRepair"`
+}
+
+// StaticNamespaceRetention sets the retention per namespace (required)
+type StaticNamespaceRetention struct {
+	RetentionPeriod                     time.Duration `yaml:"retentionPeriod"`
+	BlockSize                           time.Duration `yaml:"blockSize"`
+	BufferFuture                        time.Duration `yaml:"bufferFuture"`
+	BufferPast                          time.Duration `yaml:"bufferPast"`
+	BlockDataExpiry                     bool          `yaml:"blockDataExpiry"`
+	BlockDataExpiryAfterNotAccessPeriod time.Duration `yaml:"blockDataExpiryAfterNotAccessPeriod"`
+}
+
+// NewNamespaceInitializer creates a new namespace initializer
+func (c Configuration) NewNamespaceInitializer(
+	iopts instrument.Options,
+	configSvcClient client.Client,
+) (namespace.Initializer, error) {
+
+	switch {
+	case c.Service != nil:
+		dynamicOpts := namespace.NewDynamicOptions().
+			SetInstrumentOptions(iopts).
+			SetConfigServiceClient(configSvcClient).
+			SetNamespaceRegistryKey(kvconfig.NamespacesKey)
+		nsInit := namespace.NewDynamicInitializer(dynamicOpts)
+		return nsInit, nil
+
+	case c.Static != nil:
+		nsList := []namespace.Metadata{}
+		for _, ns := range c.Static.Namespaces {
+			md, err := newNamespaceMetadata(ns)
+			if err != nil {
+				err = fmt.Errorf("unable to create metadata for static config: %v", err)
+				return nil, err
+			}
+			nsList = append(nsList, md)
+		}
+
+		nsInitStatic := namespace.NewStaticInitializer(nsList)
+		return nsInitStatic, nil
+
+	default:
+		return nil, errMissingNamespaceConfig
+	}
+}
+
+// NewTopologyInitializer creates a new topology initializer
+func (c Configuration) NewTopologyInitializer(
```

Comment:
So on thinking this through a fair bit, we should probably just make this one method yeah.  

With respect to returning things callers don't need: the namespace initializer isn't expensive, it's just an initializer.  It doesn't acquire any resources to create.